browser/brave_vpn/dns/brave_vpn_dns_observer_service_unittest.cc

```diff
@@ -0,0 +1,517 @@
+    DisableDoHSystemConfig();
+    FireBraveVPNStateChange(mojom::ConnectionState::CONNECTED);
+    ExpectDNSMode(SecureDnsConfig::kModeAutomatic, custom_servers);
+    EXPECT_FALSE(pref_service()->GetDict(prefs::kBraveVPNUserConfig).empty());
+    EXPECT_FALSE(pref_service()->GetBoolean(prefs::kBraveVPNUserConfigLocked));
+    FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTING);
+    ExpectDNSMode(SecureDnsConfig::kModeAutomatic, custom_servers);
+    FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTED);
+    ExpectDNSMode(SecureDnsConfig::kModeAutomatic, custom_servers);
+    EXPECT_TRUE(pref_service()->GetDict(prefs::kBraveVPNUserConfig).empty());
+  }
+
+  // Browser DoH mode automatic with broken custom servers, System DoH off
+  // -> override browser config and enable vpn
+  {
+    SetDNSMode(SecureDnsConfig::kModeAutomatic, std::string());
+    FireBraveVPNStateChange(mojom::ConnectionState::CONNECTING);
+    ExpectDNSMode(SecureDnsConfig::kModeAutomatic, std::string());
+    DisableDoHSystemConfig();
+    FireBraveVPNStateChange(mojom::ConnectionState::CONNECTED);
+    ExpectDNSMode(SecureDnsConfig::kModeSecure, default_servers);
+    EXPECT_FALSE(pref_service()->GetDict(prefs::kBraveVPNUserConfig).empty());
+    EXPECT_TRUE(pref_service()->GetBoolean(prefs::kBraveVPNUserConfigLocked));
+    FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTING);
+    ExpectDNSMode(SecureDnsConfig::kModeSecure, default_servers);
+    FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTED);
+    ExpectDNSMode(SecureDnsConfig::kModeAutomatic, std::string());
+    EXPECT_FALSE(pref_service()->GetBoolean(prefs::kBraveVPNUserConfigLocked));
+    EXPECT_TRUE(pref_service()->GetDict(prefs::kBraveVPNUserConfig).empty());
+  }
+
+  // Browser DoH mode secure with custom servers, System DoH on
+  // -> do not override browser config and enable vpn
+  {
+    pref_service()->ClearPref(prefs::kBraveVPNUserConfig);
+    pref_service()->ClearPref(prefs::kBraveVPNUserConfigLocked);
+    SetDNSMode(SecureDnsConfig::kModeSecure, custom_servers);
+    FireBraveVPNStateChange(mojom::ConnectionState::CONNECTING);
+    ExpectDNSMode(SecureDnsConfig::kModeSecure, custom_servers);
+    SetValidDoHSystemConfig();
+    FireBraveVPNStateChange(mojom::ConnectionState::CONNECTED);
+    ExpectDNSMode(SecureDnsConfig::kModeSecure, custom_servers);
+    EXPECT_FALSE(pref_service()->GetDict(prefs::kBraveVPNUserConfig).empty());
+    EXPECT_FALSE(pref_service()->GetBoolean(prefs::kBraveVPNUserConfigLocked));
+    FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTING);
+    ExpectDNSMode(SecureDnsConfig::kModeSecure, custom_servers);
+    FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTED);
+    ExpectDNSMode(SecureDnsConfig::kModeSecure, custom_servers);
+    EXPECT_TRUE(pref_service()->GetDict(prefs::kBraveVPNUserConfig).empty());
+  }
+  // Browser DoH mode secure with custom servers, System DoH off
```

Comment:
system doh is not relevant, the actions we take on the browser depend only on the browser doh settings

Reply:
removed